### PR TITLE
Ocaml_vm

### PR DESCRIPTION
**Pull request by Zett98**
## Problem

We dont have a wasm vm written in ocaml on Deku

## Solution

Add it

### CODE REVIEW

File: deku-p/src/core/protocol/protocol.ml [hunk 1]
```diff
@@ -112,6 +112,7 @@ let apply_operation ~current_level protocol operation :
               with
               | vm_state, ledger -> Some (ledger, Some receipt, vm_state, None)
               | exception External_vm_client.Vm_execution_error error ->
+                  let () = failwith error in
```
Comment (d4hines) [inline deku-p/src/core/protocol/protocol.ml:115]: This doesn't look right.

**Review by d4hines — CHANGES_REQUESTED** [review]

**Review by EduardoRFS — CHANGES_REQUESTED** [review]
Why did you vendor the interpreter? I assume it is because you made changes, if so which changes? Clean the git history.

**Comment by EduardoRFS** [comment]
Can we go without the WASM changes for now? Specially without the unicode one's, they're tricky to review and I would lilke more time on those.

**Comment by EduardoRFS** [comment]
Also @renatoalencar please review it

File: deku-c/wasm-vm-ocaml/env.mlid [hunk 2]
```diff
@@ -0,0 +1,15 @@
+(* module Contract_address = Deku_ledger.Contract_address
+   module Address = Deku_ledger.Address
+   module Ledger = Deku_ledger.Ledger
+
```
Comment (EduardoRFS) [inline deku-c/wasm-vm-ocaml/env.mlid:4]: Delete this file?

File: deku-p/src/core/ledger/ticket_id.mli [hunk 3]
```diff
@@ -4,7 +4,7 @@ type ticketer =
 [@@deriving eq, ord, show]
 
 type ticket_id = private Ticket_id of { ticketer : ticketer; data : bytes }
-and t = ticket_id [@@deriving ord, yojson, show]
+and t = ticket_id [@@deriving ord, eq, yojson, show]
```
Comment (EduardoRFS) [inline deku-p/src/core/ledger/ticket_id.mli:7]: Why do you need this eq?

File: deku-c/wasm-vm-ocaml/value.mli [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  and t =
+    | Int of Z.t
+    | String of string
+    | Bytes of bytes
+    | Bool of int
+    | Pair of t * t
+    | Union of union
+    | List of t list * typ
+    | Option of t option
+    | Unit
```
Comment (EduardoRFS) [inline deku-c/wasm-vm-ocaml/value.mli:18]: Unit should be on tuna

**Comment by EduardoRFS** [comment]
Those changes are okay, most of the code is gonna die, so if possible to remove the changes to the interpreter, let's go with it.

File: deku-c/wasm-vm-ocaml/env.mlid (hunk 2 above)
Comment (Zett98) [reply deku-c/wasm-vm-ocaml/env.mlid]: updated the mli

File: deku-p/src/core/ledger/ticket_id.mli (hunk 3 above)
Comment (Zett98) [reply deku-p/src/core/ledger/ticket_id.mli]: to not do polymorphic eq, but the current vm will be dropped anyways, so i'll remove the eq derivation for now

**Comment by Zett98** [comment]
> Those changes are okay, most of the code is gonna die, so if possible to remove the changes to the interpreter, let's go with it.

I'd keep the changes for now, as they simplified gas metering and decoding and encoding of idents.
With regards to utf-8, Does it even matter? Vm will be rewritten soon enough, and it works with current contracts

File: deku-p/src/core/protocol/protocol.ml (hunk 1 above)
Comment (d4hines) [reply deku-p/src/core/protocol/protocol.ml:115]: Pavlo explained that it's fixed in the next PR.

**Comment by d4hines** [comment]
From my POV it would be better to merge https://github.com/marigold-dev/deku/pull/954 into this one and merge them together atomically. Other than that LGTM

**Comment by EduardoRFS** [comment]
From my POV this is bad, but it is an improvement over whatever we have. So let's go with it.

File: deku-c/wasm-vm-ocaml/operation.ml [hunk 5]
```diff
@@ -0,0 +1,17 @@
+open Deku_ledger
+open Deku_concepts
+
+type t = {
+  operation : Operations.t;
```
Comment (renatoalencar) [inline deku-c/wasm-vm-ocaml/operation.ml:5]: I don't see why there's an Operation and an Operations module and how operations could more than one thing, the names should probably change in order to settle the confusion.

File: deku-c/wasm-vm-ocaml/env.ml [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+                         in
+                         ignore @@ Wasm.Eval.invoke func [ arg ]);
+                    Some (fun (k : (a, _) continuation) -> continue k ())
+                | _ -> None);
+          }
+    with
+    | Failure s -> Error ("unexpected failure. Failure " ^ s)
+    | Wasm.Eval.Crash _ | Wasm.Eval.Exhaustion _ | Wasm.Eval.Link _
+    | Wasm.Eval.Trap _ ->
+        Error "Invalid contract supplied to the vm"
```
Comment (renatoalencar) [inline deku-c/wasm-vm-ocaml/env.ml:173]: Don't ignore the actual errors, that makes debugging harder. I think something similar to line 53 is enough for now.

**Review by renatoalencar — CHANGES_REQUESTED** [review]
Looks fine in general, just two small things I've noticed.

**Review by renatoalencar — APPROVED** [review]
Looks good

**Review by d4hines — APPROVED** [review]
:+1:

**Pull request merged**